### PR TITLE
fix(oid4vp): complete SD-JWT DCQL query-criteria validation for VP token

### PR DESCRIPTION
**Pull request by mbunwe-victor**
## Summary

This PR completes OpenID4VP 1.0 §8.6(2.2) DCQL query-criteria validation for the SD-JWT VC login flow.

Previously, VP token validation only checked whether requested claim paths were present. This PR validates full DCQL criteria satisfaction during presentation validation.

### What changed

- Resolve nested string object-key claim paths, including through selectively disclosed SD-JWT claims (e.g. `["address", "street_address"]`)
- Validate DCQL `claim.values` during presentation validation; reject empty `values` arrays at query build time
- Apply runtime `claim_sets` satisfaction (at least one option must match)
- Keep `Claim.path` and `Claim.values` as `List<String>`, matching the plugin’s current DCQL query builder
- Keep existing SD-JWT VC scope (single credential, `dc+sd-jwt`)

## Test plan

- [x] `DcqlQueryValidatorTest`

### CODE REVIEW

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/ClaimPathResolver.java [hunk 1]
```diff
@@ -2,66 +2,137 @@
...
-            return hasDisclosedClaim(sdJwt, path.getFirst());
+        if (path.size() == 1 && path.getFirst() instanceof String claimName) {
+            JsonNode disclosedClaim = findDisclosedClaim(sdJwt, claimName);
+            if (isPresent(disclosedClaim)) {
+                return List.of(disclosedClaim);
+            }
         }
 
-        return false;
+        return List.of();
```
Comment (forkimenjeckayang) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/ClaimPathResolver.java:36]: The resolver only falls back to SD-JWT disclosures when the requested path has exactly one segment. That means a valid DCQL claim such as `["address", "street_address"]` fails whenever `address` is selectively disclosed as an object, because the issuer payload lookup misses it and the disclosure lookup is skipped for paths longer than one segment. OpenID4VP Final 1.0 Section 7.4 explicitly shows `["address", "street_address"]` for `dc+sd-jwt`, and Section 7 says the pointer is evaluated against the credential, not only against clear-text issuer payload claims. Please reuse/extend the existing disclosure-reading approach so `resolveInSdJwt` can resolve the first disclosed claim value and then apply the remaining path segments with `resolveInJson`.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/model/dcql/Claim.java [hunk 2]
```diff
@@ -8,15 +8,11 @@ public class Claim {
...
-     * Object-key path segments only (e.g. {@code credentialSubject}, {@code given_name}). Does not
-     * support {@code null} array wildcards or integer indexes from OpenID4VP Claims Path Pointers.
-     */
     @JsonProperty("path")
-    private List<String> path;
+    private List<Object> path;
 
     @JsonProperty("values")
-    private List<String> values;
+    private List<Object> values;
```
Comment (forkimenjeckayang) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/model/dcql/Claim.java]: This PR widens `Claim.values` to `List<Object>` so integer and boolean constraints can work, but `DcqlQueryValidator` never validates the `values` array. Per OpenID4VP Final 1.0 Section 6.3, `values`, when present, must be a non-empty array of allowed literal values: string, integer, or boolean. With the current code, `values: []` is silently treated as no restriction at runtime, and object/array/null values can be accepted and compared through `valueToTree`, which distorts the DCQL query model. Please add query validation for non-empty `values` and reject anything other than string, integer, or boolean, preferably near the existing claim path validation so malformed DCQL is rejected before presentation validation.

Related file: `src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlQueryValidator.java`
- Related changed lines: 160-197

**Review by forkimenjeckayang — DISMISSED** [review]
a few comments to check @mbunwe-victor

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/model/dcql/Claim.java
Comment (Ogenbertrand) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/model/dcql/Claim.java:1]: There's no need changing String to Object, please read the comment here and understand: https://github.com/ADORSYS-GIS/keycloak-oid4vp-plugin/pull/94/changes#r3361700505

Let's continue the review after that. I'll also appreciate @IngridPuppet review here.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlPresentationValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                claimSet -> claimSet.put("given_name", "Alice"),
+                DisclosureSpec.builder().build());
+
+        DcqlQuery query = queryWithClaims(claimWithValues("given-name", List.of("given_name"), List.of("Bob")));
+
+        assertThrows(
+                VerificationException.class,
+                () -> DcqlPresentationValidator.validatePresentation(query, token),
+                "Should reject presentation when claim values do not match");
+    }
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlPresentationValidatorTest.java]: Assert the error message.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlPresentationValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        Credential credential = credentialWithClaims(List.of(givenName, familyName));
+        credential.setClaimSets(List.of(List.of("given-name", "family-name")));
+
+        DcqlQuery query = DcqlQueryBuilder.singleCredentialQuery(credential);
+
+        assertThrows(
+                VerificationException.class,
+                () -> DcqlPresentationValidator.validatePresentation(query, token),
+                "Should reject presentation when no claim_sets option is satisfied");
+    }
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlPresentationValidatorTest.java]: Assert the error message.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlQueryValidatorTest.java
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/DcqlQueryValidatorTest.java:1]: Assert error messages across negative tests.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/ClaimPathResolver.java [hunk 5]
```diff
@@ -2,66 +2,98 @@
...
     private ClaimPathResolver() {}
 
     public static boolean isPresentInJson(JsonNode root, List<String> path) {
-        return isPresent(resolveInPayload(root, path));
+        return !resolveInJson(root, path).isEmpty();
     }
 
     public static boolean isPresentInSdJwt(SdJwtVP sdJwt, List<String> path) {
+        return !resolveInSdJwt(sdJwt, path).isEmpty();
+    }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/dcql/ClaimPathResolver.java]: These methods are reported as unused.

**Review by IngridPuppet — COMMENTED** [review]
Hello @mbunwe-victor - Thank you for the PR. I left some comments below. I also think the test coverage could be improved to cover more negative branches. Please could you work on that?

Image

**Comment by mbunwe-victor** [comment]
Test coverage status:

image

**Review by IngridPuppet — APPROVED** [review]

**Review by Ogenbertrand — APPROVED** [review]
LGTM
